src/element/element_stream_writer.rs

```diff
@@ -0,0 +1,697 @@
+        writer.flush().expect("Call to flush() failed.");
+        let expected: Vec<Element> = expected.into_iter().map(|e| e.into()).collect();
+
+        assertion(output, expected);
+    }
+
+    #[track_caller]
+    fn writer_test<F, E>(commands: F, expected: Vec<E>)
+    where
+        E: Into<Element>,
+        F: FnMut(&mut ElementStreamWriter<Vec<Element>>) -> IonResult<()>,
+    {
+        writer_test_with_assertion(commands, expected, |actual, expected| {
+            assert_eq!(actual, expected);
+        })
+    }
+
+    #[track_caller]
+    fn writer_scalar_test_with_assertion<F, E, A>(commands: F, expected: E, mut assertion: A)
+    where
+        E: Into<Element>,
+        F: FnMut(&mut ElementStreamWriter<Vec<Element>>) -> IonResult<()>,
+        A: FnMut(Element, Element),
+    {
+        writer_test_with_assertion(
+            commands,
+            vec![expected.into()],
+            |mut actual, mut expected| {
+                assert_eq!(actual.len(), 1);
+                assert_eq!(expected.len(), 1);
+                assertion(actual.pop().unwrap(), expected.pop().unwrap());
+            },
+        );
+    }
+
+    #[track_caller]
+    fn writer_scalar_test<F, E>(commands: F, expected: E)
+    where
+        E: Into<Element>,
+        F: FnMut(&mut ElementStreamWriter<Vec<Element>>) -> IonResult<()>,
+    {
+        writer_scalar_test_with_assertion(commands, expected, |actual, expected| {
+            assert_eq!(actual, expected);
+        })
+    }
+
+    #[test]
+    fn write_null_null() {
+        writer_scalar_test(|w| w.write_null(IonType::Null), Value::Null(IonType::Null));
+    }
+
+    #[test]
+    fn write_null_string() {
+        writer_scalar_test(
+            |w| w.write_null(IonType::String),
+            Value::Null(IonType::String),
+        );
+    }
+
+    #[test]
+    fn write_bool_true() {
+        writer_scalar_test(|w| w.write_bool(true), Value::Bool(true));
+    }
+
+    #[test]
+    fn write_bool_false() {
+        writer_scalar_test(|w| w.write_bool(false), Value::Bool(false));
+    }
+
+    #[test]
+    fn write_i64() {
+        writer_scalar_test(|w| w.write_i64(7), Value::Int(Int::I64(7)));
+    }
+
+    #[test]
+    fn write_f32() {
+        writer_scalar_test(|w| w.write_f32(700f32), Value::Float(700.));
+    }
+
+    #[test]
+    fn write_f64() {
+        writer_scalar_test(|w| w.write_f64(700f64), Value::Float(700.));
+    }
+
+    #[test]
+    fn write_annotated_i64() {
+        writer_scalar_test(
+            |w| {
+                w.set_annotations(["foo", "bar", "baz quux"]);
+                w.write_i64(7)
+            },
+            Value::Int(Int::I64(7)).with_annotations(["foo", "bar", "baz quux"]),
+        );
+    }
+
+    #[test]
+    fn write_decimal() {
+        // returns (BigInt, BigInt) representing the differences between integer and fractional
+        // components of d1 & d2 respectively
+        fn parts_diff(d1: Decimal, d2: Decimal) -> (BigInt, BigInt) {
```

Comment:
This is copied verbatim from #562 . Perhaps it could be extracted somewhere appropriate?

Reply:
Yeah, I think this can live in `Decimal`. Could you add a couple of example input/output pairs to the doc comment to illustrate its behavior?

Reply:
Moved in #562, will rebase this onto that and use it here.